plugins/modem-manager/fu-qmi-pdc-updater.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (C) 2019 Aleksander Morgado <aleksander@aleksander.es>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-qmi-pdc-updater.h"
+
+#define FU_QMI_PDC_MAX_OPEN_ATTEMPTS 8
+
+struct _FuQmiPdcUpdater {
+	GObject		parent_instance;
+	gchar		*qmi_port;
+	QmiDevice	*qmi_device;
+	QmiClientPdc	*qmi_client;
+};
+
+G_DEFINE_TYPE (FuQmiPdcUpdater, fu_qmi_pdc_updater, G_TYPE_OBJECT)
+
+typedef struct {
+	GMainLoop	*mainloop;
+	QmiDevice	*qmi_device;
+	QmiClientPdc	*qmi_client;
+	GError		*error;
+	guint		open_attempts;
+} OpenContext;
+
+static void fu_qmi_pdc_updater_qmi_device_open_attempt (OpenContext *ctx);
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_abort_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	g_assert (ctx->error != NULL);
```

Comment:
:/ -- g_critical?

Reply:
So this is the thing, error will always be set, because this code path is the "abort" path, and the error set in the context was the last error we got before we started the abort path. Do you still prefer a critical warning in this case? Otherwise, maybe just removing the assert is better? From my point of view, if the next logic relies on that error being set, the assert makes sense, but really not sure how you handle these cases in fwupd.

Reply:
Just remove the assert I guess.